File: src/main/java/com/github/svegon/capi/event/world/ClientWorldLifecycleEvents.java
package com.github.svegon.capi.event.world;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.client.world.ClientWorld;

public final class ClientWorldLifecycleEvents {
    private ClientWorldLifecycleEvents() {
        throw new UnsupportedOperationException();
    }

    public static final Event<JoinWorld> JOIN_WORLD = EventFactory.createArrayBacked(JoinWorld.class,
            listeners -> (client, world) -> {
                for (JoinWorld listener : listeners) {
                    listener.onWorldJoin(client, world);
                }
            });

    public static final Event<LeaveWorld> LEAVE_WORLD = EventFactory.createArrayBacked(LeaveWorld.class,
            listeners -> (client, screen) -> {
                for (LeaveWorld listener : listeners) {
                    listener.onWorldLeave(client, screen);
                }
            });

    public interface JoinWorld {
        void onWorldJoin(MinecraftClient client, ClientWorld world);
    }

    public interface LeaveWorld {
        void onWorldLeave(MinecraftClient client, Screen screen);
    }
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/world/ClientWorldLifecycleEvents.kt
package io.github.svegon.capi.event.world

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.client.MinecraftClient
import java.util.function.Function

class ClientWorldLifecycleEvents private constructor() {
    init {
        throw UnsupportedOperationException()
    }

    interface JoinWorld {
        fun onWorldJoin(client: MinecraftClient?, world: ClientWorld?)
    }

    interface LeaveWorld {
        fun onWorldLeave(client: MinecraftClient?, screen: Screen?)
    }

    companion object {
        @JvmField
        val JOIN_WORLD: Event<JoinWorld> = EventFactory.createArrayBacked<JoinWorld>(
            JoinWorld::class.java,
            Function<Array<JoinWorld>, JoinWorld> { listeners: Array<JoinWorld> ->
                JoinWorld { client: MinecraftClient?, world: ClientWorld? ->
                    for (listener in listeners) {
                        listener.onWorldJoin(client, world)
                    }
                }
            })

        @JvmField
        val LEAVE_WORLD: Event<LeaveWorld> = EventFactory.createArrayBacked<LeaveWorld>(
            LeaveWorld::class.java,
            Function<Array<LeaveWorld>, LeaveWorld> { listeners: Array<LeaveWorld> ->
                LeaveWorld { client: MinecraftClient?, screen: Screen? ->
                    for (listener in listeners) {
                        listener.onWorldLeave(client, screen)
                    }
                }
            })
    }
}
